#### [**Mathematical Modeling of Unsteady Inviscid Flows**](https://www.springer.com/gp/book/9783030183189)
**by Jeff D. Eldredge** (Springer, 2019)

This Jupyter notebook and associated code serve as a companion to the book. The notebook is powered by the [`PotentialFlow`](https://github.com/darwindarak/PotentialFlow.jl) package, written in the [Julia language](https://julialang.org/) by Darwin Darakananda and Jeff D. Eldredge, UCLA. *The code is released under the [MIT license](https://opensource.org/licenses/MIT).*
<hr />

<!--NAVIGATION-->
< [Contents](Index.ipynb) | [Streamfunction of rigid-body motion](3.1-StreamfunctionOfRigidBody.ipynb) >

<a id='top'></a>
## Introduction

#### Some basics of the Julia language

These notebooks are written in the [Julia language](https://julialang.org/), which has its own set of excellent documentation and examples. For the most part, the notebooks that follow are self-explanatory. However, here we will provide some basic skills to allow you to navigate through these notebooks.

In many cases, code is embedded directly into the cells of the notebook. You can then execute this code simply by running the cell. The easiest way to do this is to press `<shift>`-`<enter>`, or to press the Run (right-arrow) button on the bar above. Try it here:

In [1]:
2+2

4

In defining variables, functions, etc., Julia accepts *unicode characters*, including various symbols, Greek letters, and sub/superscripts in addition to the usual set of numbers and Latin letters. To obtain these symbols, use the backslash (i.e., escape) key. For example, to get $\alpha$, type `\alpha` and then `<tab>` to execute it:

In [2]:
α = 1

1

In particular, $\pi$ is reserved for the usual irrational number:

In [3]:
π

π = 3.1415926535897...

Also note that the imaginary unit $i$ is written in Julia as `im`

In [4]:
im

im

For example, we can form a complex number $2+3i$ with

In [5]:
z = 2.0+3.0im

2.0 + 3.0im

The *type* of this number is different from real-valued numbers:

In [6]:
typeof(z)

Complex{Float64}

In [7]:
x = 1.0
typeof(x)

Float64

A subscript is generated with `\_` followed by the desired subscript character:

In [8]:
x₃ = 1

1

If you wish to suppress the output of a cell when you execute it, then put a semicolon `;` at the end:

In [9]:
x = 2;

The easiest way to set up a uniform 1d grid of numbers is with the `range` function:

In [10]:
x = range(-1,1,length=201)

-1.0:0.01:1.0

If you wish to learn more about the syntax of a Julia function, then write `?` followed by the function name to get help:

In [11]:
?range

search: range LinRange UnitRange StepRange StepRangeLen trailing_zeros



```
range(start[, stop]; length, stop, step=1)
```

Given a starting value, construct a range either by length or from `start` to `stop`, optionally with a given step (defaults to 1, a [`UnitRange`](@ref)). One of `length` or `stop` is required.  If `length`, `stop`, and `step` are all specified, they must agree.

If `length` and `stop` are provided and `step` is not, the step size will be computed automatically such that there are `length` linearly spaced elements in the range (a [`LinRange`](@ref)).

If `step` and `stop` are provided and `length` is not, the overall range length will be computed automatically such that the elements are `step` spaced (a [`StepRange`](@ref)).

`stop` may be specified as either a positional or keyword argument.

!!! compat "Julia 1.1"
    `stop` as a positional argument requires at least Julia 1.1.


# Examples

```jldoctest
julia> range(1, length=100)
1:100

julia> range(1, stop=100)
1:100

julia> range(1, step=5, length=100)
1:5:496

julia> range(1, step=5, stop=100)
1:5:96

julia> range(1, 10, length=101)
1.0:0.09:10.0

julia> range(1, 100, step=5)
1:5:96
```


Any element in this array can be accessed with square brackets:

In [12]:
x[5]

-0.96

Actually, `x` is not a column vector yet, but rather, a range of values that can be interpreted as an array when necessary. To get the actual array with all of the values, use `collect`:

In [13]:
y = collect(x)

201-element Array{Float64,1}:
 -1.0 
 -0.99
 -0.98
 -0.97
 -0.96
 -0.95
 -0.94
 -0.93
 -0.92
 -0.91
 -0.9 
 -0.89
 -0.88
  ⋮   
  0.89
  0.9 
  0.91
  0.92
  0.93
  0.94
  0.95
  0.96
  0.97
  0.98
  0.99
  1.0 

1d arrays are, by default, interpreted as column vectors, but they can be transposed into row vectors with the `'`:

In [14]:
y'

1×201 LinearAlgebra.Adjoint{Float64,Array{Float64,1}}:
 -1.0  -0.99  -0.98  -0.97  -0.96  …  0.95  0.96  0.97  0.98  0.99  1.0

Vectorized operations (i.e, operations carried out element by element) can be applied with the `.` notation. For example, to add `1` to every entry in `x`:

In [15]:
x .+ 1

0.0:0.01:2.0

If we wish to assign an array to a new variable, and that variable has already been initialized (or *instantiated*, as they say), then a good memory and time saver is to use `.=` to reassign the variable's elements.

In [16]:
y = zeros(length(x))

201-element Array{Float64,1}:
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 ⋮  
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0

In [17]:
y .= x .+ 1;

This reassigned all of the entries in `y` to the value of `x + 1`, element by element. To see how efficient that operation was, use the `@time` macro:

In [18]:
@time y .= x .+ 1;

  0.000008 seconds (6 allocations: 288 bytes)


Only 6 allocations of 288 bytes. If we had not used the `.=`, it would have been much more. For example:

In [19]:
@time z = collect(x) .+ 1;

  0.057931 seconds (211.16 k allocations: 10.844 MiB)


Considerably more memory was allocated in this case. This is fine if `z` did not already exist, but if `z` does already exist, it is wasteful to reallocate memory for it.

Other Julia syntax will be introduced as it arises in the notebooks that follow.

[Return to top of notebook](#top)

<!--NAVIGATION-->
< [Contents](Index.ipynb) | [Streamfunction of rigid-body motion](3.1-StreamfunctionOfRigidBody.ipynb) >